Train MainBTB & TAGE predictor at resolve stage

Yakkhini opened the pull request.

Yakkhini committed:
cpu-o3: split bpu update and commit interface in btb and fetch

Change-Id: Ie683a93f083c35afb6578e1b039063bd9061f41a

Yakkhini committed:
cpu-o3: record resolved FSQ Id and send back to fetch stage

Change-Id: Iaec9f9712d0b8156340f153accb163a6be2e56e6

Yakkhini committed:
cpu-o3: train predictors via resolved FSQ id in fetch

Change-Id: Idecec2b0a2d2d2d821dfef78d15a9a98619c70c2

Yakkhini committed:
cpu-o3: only train TAGE in resolve update

Change-Id: I19e8d087ede8cd6fdfd8393bdc5b3d36fe4a68f8

Yakkhini committed:
cpu-o3: push resolved inst pc with FSQ id for btb entry matching

Change-Id: I65c3224ef153d7bbb873bf7fc926283a7457d0d2

Yakkhini committed:
cpu-o3: add resolved bit in BTB entry for TAGE training skip

Change-Id: I6d821221f475cf7cd931f9dcc13a21255d4bdd39

Yakkhini committed:
cpu-o3: btb resolved and tage trained signal to avoid multi train

Change-Id: I124b95c5854301aca6cccfb6a856aa31ec6f1713

Yakkhini committed:
cpu-o3: mark resolved btb entry and train tage in fetch

Change-Id: Ia188dcb1498504a95b02ad1120e241752a1d6422

Yakkhini committed:
cpu-o3: not train predictors when squash accur in btb

Change-Id: I1aeaa4881eb4a1f9820b457e565ec4a040099fe2

Yakkhini committed:
cpu-o3: set update btb entries before mark CFI resolved

Change-Id: I41bd9f7b8c0d8b8d13bae7c3a1f11086426534d9

Yakkhini committed:
cpu-o3: mark new potential entries as resolved

Change-Id: I1736e4461292468482f3cd0d651d54e0bacb29e1

Yakkhini committed:
cpu-o3: train TAGE when no squash & no commit stream

Change-Id: I33085b2f127d2a36dfe93b0f0f4ae66e92546e2b

Yakkhini committed:
cpu-o3: adjust I2F delay to suqash before tage resolve train

This will also affect ldstqCount. Does it matter?

Change-Id: I4f1729d411deebeadd88fbbbb3dba30853487a30

Yakkhini committed:
cpu-o3: prepare resolve entries once to avoid clear after mark

In this case, no need to use `tageTrained` field to mark btb entries
since it will clear resolved value every time.

Change-Id: Ib75a7a6454df84464d98a289426903a2b6da2e24

Yakkhini committed:
cpu-o3: set new btb entries resolved default to false

Change-Id: Iadbb3611ea63bb0b1bd630ef5ae64df6e36d75ec

Yakkhini committed:
cpu-o3: wrap tage training process into iew signal handler

Change-Id: I007fac7f1608be8e8e7abb57bf4029f3a51b7344

Yakkhini committed:
cpu-o3: not check resolved btb entry when train microtage

Change-Id: Ia0136eb72115c4ca340e05eb08fe3bd9316d9dc3

Yakkhini committed:
cpu-o3: use potential new entry to resolve train tage

Change-Id: I895a1dfcf4a007c4f94b61313b6de8dc3e7bdfab

Yakkhini committed:
cpu-o3: follow btb -> mbtb as API change in 072fe6c9

Change-Id: I15bcdde28f5609c447c76e984332a785dc4eaefc

Yakkhini committed:
cpu-o3: only prepare btb entries when need

Change-Id: Ica239468a741ca9533d25352b1776e64d04c661f

Yakkhini committed:
cpu-o3: also train MainBTB in resolve stage

Change-Id: I4fcdc3d1661825c14af06e7f5383c240b8dc23cb

Yakkhini committed:
cpu-o3: match resolved pc and fsq id to avoid unexpected training

Change-Id: Ief6b17aa2f9530ec903d9826a8ca99528301fcdd

Yakkhini committed:
cpu-o3: disable resolve update by default in ci

Change-Id: I5307550203479f661c52b532b48a85e49cadf3b0

Yakkhini commented:
> ## 🚀 Coremark Smoke Test Results
> Branch 	IPC 	Change
> Base (`xs-dev`) 	`2.1375` 	-
> This PR 	`2.0856` 	📉 `-0.0519` (`-2.43%`)
> 
> ✅ Difftest smoke test passed!

According to local coremark running and analyze, MKPI is reduced when only train up to single resolve queue entry per cycle, but there is more performance decrease in non-frontend bound.

Yakkhini committed:
cpu-o3: fix python config typo xiangsha -> xiangshan

Change-Id: Icef31061b720547324ee937121283dab5e53d55d

jensen-yan commented on src/cpu/pred/btb/btb_tage.cc:
```diff
@@ -75,6 +76,7 @@ useAltOnNaWidth(p.useAltOnNaWidth),
 numTablesToAlloc(p.numTablesToAlloc),
 enableSC(p.enableSC),
 updateOnRead(p.updateOnRead),
+resolvedUpdate(false),
```
why not use p.resolvedUpdate, set in BranchPredictor.py

jensen-yan commented on src/cpu/o3/fetch.cc:
```diff
@@ -1503,6 +1528,7 @@ Fetch::handleCommitSignals(ThreadID tid)
             } else if (isBTBPred()) {
                 assert(dbpbtb);
+                dbpbtb->commitStream(fromCommit->commitInfo[tid].doneFsqId);
                 dbpbtb->update(fromCommit->commitInfo[tid].doneFsqId, tid);
```
use commitStream instead of update, bpu.update is not used,
Either delete it, or let's not rename it for now, just add some comments to the update function.
This doesn't feel very elegant.

jensen-yan commented on src/cpu/o3/iew.cc:
```diff
@@ -1656,6 +1673,10 @@ IEW::executeInsts()
 
+    // Clear resolvedFSQId and resolvedInstPC since they are already handled in frontend
+    ThreadID tid = *activeThreads->begin();
+    toFetch->iewInfo[tid].resolveQueue.clear();
```
The `resolveQueue` is currently a member of the forward pipeline, meaning a new queue is created for each iteration, and all processes in that iteration are completed. This is actually slightly different from the persistent queue implementation in RTL.

Yakkhini replied:
Will align to RTL in future

Yakkhini committed:
cpu-o3: use predictor options to trigger resolve update feature

Change-Id: I3914c0195f788af17764b92594bff2108bf6e0e5

Yakkhini committed:
cpu-o3: move resolve update options to basic predictor class

Just for better code style and future extensibility. Currently support
MainBTB and TAGE only.

Change-Id: Ib2efebec847cf0c15bac09940ab3c4e3c9a75593

Yakkhini committed:
cpu-o3: disable resolve update since still no perf benefits now

Change-Id: Iead74a3c5e04a6e45d9494491154cbba4bfeedc1

Yakkhini committed:
cpu-o3: update functions style and comments for merge

Change-Id: Ie6e2895fbfdd6986e3a85b2a95ff980fbf3d175f

Yakkhini commented on src/cpu/pred/btb/decoupled_bpred.cc:
```diff
@@ -1170,7 +1238,9 @@ DecoupledBPUWithBTB::updatePredictorComponents(FetchStream &stream)
         // Update all predictor components
         for (int i = 0; i < numComponents; ++i) {
-            components[i]->update(stream);
+            if (components[i] != tage || components[i] != mbtb) {
```
That's wrong and the code is refactored now.
